[IMP] project: tasks dependencies copy when generating next occurrence of recurring task

deneuvillem opened the pull request:
Before this commit, when a recurring task with sub-tasks and dependencies was re-generated due to the next occurrence of the recurrence, the dependencies were lost in the newly created task.

After this commit, the dependency structure is kept the same in the newly created task as in the original one.
Moreover, the next occurrence generation has been simplified and is now using the copy() method.

task-3716523

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/152916).

damrOdoo commented on addons/project/models/project_task_recurrence.py:
```diff
@@ -79,11 +79,28 @@ def _get_recurrence_delta(self):
 
+    def _copy_task_dependency(self, task_copy, task_original):
+        if 'task_mapping' not in self.env.context:
+            self = self.with_context(task_mapping=dict())
```
i'm really not a big fan of using the context like that to build the dict. It would be better to pass an empty dict as an argument, and slowly build it up with the return value of each function.

damrOdoo commented on addons/project/models/project_task_recurrence.py:
```diff
@@ -79,11 +79,28 @@ def _get_recurrence_delta(self):
+        task_mapping = self.env.context.get('task_mapping')
+        task_mapping[task_original.id] = task_copy.id
+        tasks_copied = task_mapping.values()
+        task_original.write({'depend_on_ids': [Command.unlink(t.id) for t in task_original.depend_on_ids if t.id in tasks_copied]})
+        task_original.write({'dependent_ids': [Command.unlink(t.id) for t in task_original.dependent_ids if t.id in tasks_copied]})
+        task_copy.write({'depend_on_ids': [Command.link(task_mapping.get(t.id, t.id)) for t in task_original.depend_on_ids]})
+        task_copy.write({'dependent_ids': [Command.link(task_mapping.get(t.id, t.id)) for t in task_original.dependent_ids]})
```
linking and unlinking relation like that is not very efficient. It would be better to build a dict containing the necessary values and link it in one go.

damrOdoo replied:
There are 2 things that i'm not found of here.
1) using write is usually not the best practice. it's better to simply use task.copy.depends_on = new_value instead. That way, we let the orm optimize the data updates. Unless we have a reason to specifically use 'write' in this case.
2) the fact that you have to 'unlink' some dependence on the original task, means that at some point you're linking things that are not supposed to be linked. We should try to simply not make those useless link in the first place, instead of trying to fix our 'mistake' later on.

damrOdoo reviewed (commented):
hello @deneuvillem. Thanks for your work. Unfortunately, i disagree with the way you are handling it. It is not very efficient, as well as kinda counter intuitive on first read. (it took me a while to understand why you had to do unlinks on original task :sweat_smile: )
In case like this, i think its better to create a dict instead that will contains the values we want to assign, and use these values at once.
What i had in mind was that instead of handling each child task directly, we could have the copy_task_dependency_values return a tuple containing the 2 following dict :
1) dict (key : original_task_id: copied_task_id)
2) dict (copied_task_id: depend_on_ids)

We then use these values in the super() and slowly build a tuple which contains  all the dependencies we have to copy. We can then check the ids for each dependency, and if one of them is from the original_task, then switch it  with the values of the 1st dict. once the dependencies dict is cleaned, we can add all the dependencies at once, without having to worry about removing unnecessary ones.
Do you understand what i'm suggesting, and if yes, what do you think ?

deneuvillem commented:
Hello @damrOdoo and thank you for your very detailed review and suggestions :)

I updated the code by renaming the recursive function to `_build_task_mapping_dependencies()`, which returns 3 mappings:
1. dict (key : original_task_id: copied_task_id)
2. dict (key: original_task_id: depend_on_ids)
3. dict (key: original_task_id: dependent_ids)

Then in the `_create_next_occurrence()`, we iterate over the key, value pairs of those mappings and set the dependencies of the copied tasks.
Note that I don't use the `task_mapping` to build `depend_on_mapping` and  `dependent_mapping` directly in `_build_task_mapping_dependencies()` because there is no guarantee that `task_mapping` will contain all the nodes needed by the dependency mappings at the time of the iteration. So we wait that all nodes are in the `task_mapping` before proceeding to use that mapping to clean dependencies.

Now I wonder if we couldn't just store the tasks records in `task_mapping` so that we could get rid of `depend_on_mapping` and `dependent_mapping` as we would have every information we need thanks to the records. But maybe it would be inefficient in terms of space memory to store the records inside a dictionary.

I also added the test you mentioned by adding two "side tasks" ("random tasks") that are not copied from the recurrence and check that they still appear in the new occurrence of the task.

damrOdoo commented:
Hello @deneuvillem, tbh, I dont think using a dict with record as key: values instead of ids is that heavy memory wise :thinking:. So you can try what you suggested by using the record and only one dict instead of the 3 dicts :muscle:. However, i'd suggest doing it in a second commit, that way, if the person who's doing the RD validation at the end of the process think it's not a good idea, you'll still have your draft of the ids version.

damrOdoo reviewed (commented):
hello @deneuvillem Thanks for your work. its lgtm.
just for your information, I don't like using the context like that, to pass an argument from one function to another that can be done, but its a last resort. Sadly, I did not find a better working solution.

auon-odoo reviewed (changes requested):
Thanks for your work :fire:

damrOdoo commented on addons/project/models/project_task_recurrence.py:
```diff
@@ -102,8 +82,9 @@ def _create_next_occurrence_values(self, occurrence_from):
         create_values['stage_id'] = occurrence_from.project_id.type_ids[0].id if occurrence_from.project_id.type_ids else occurrence_from.stage_id.id
         create_values['child_ids'] = [
-            Command.create(self._create_next_occurrence_values(child)) for child in occurrence_from.with_context(active_test=False).child_ids
+            child.with_context(copy_project=True).sudo().copy(self._create_next_occurrence_values(child)).id for child in occurrence_from.with_context(active_test=False).child_ids
```
I don't really like this line, since you're handling the child hierarchy here instead of letting the copy() method do it. But on the other hand, since every child has potentially custom data compared to its original version, I don't really have a better solution to offer. :man_shrugging:

xavierbol replied:
Me neither. 😕

damrOdoo reviewed (commented):
Thanks for your work. Looks good to me.

xavierbol reviewed (commented):
LGTM, thanks for your work. 🙂 

robodoo r+

robodoo commented:
@deneuvillem @xavierbol unable to stage: merge conflict

xavierbol commented:
robodoo r+
